core/include/gap/core/memoize.hpp

```diff
@@ -0,0 +1,173 @@
+// Copyright 2022-present, Trail of Bits, Inc.
+
+#pragma once
+
+#include <gap/core/concepts.hpp>
+#include <gap/core/hash.hpp>
+
+#include <tuple>
+#include <unordered_map>
+#include <utility>
+
+namespace gap
+{
+    namespace detail
+    {
+        template< typename T, typename enable_t = void >
+        struct function_signature;
```

Comment:
Pull this out into some `type_utils` header? Sounds it may be useful in other places as well.

Reply:
resolved in 46c925ccf574e5203dfd126106a679145238f102